src/notebooksyncagent.cpp

```diff
@@ -264,11 +267,70 @@ void NotebookSyncAgent::fetchRemoteChanges(const QDateTime &fromDateTime, const
     report->getAllETags(mServerPath, fromDateTime, toDateTime);
 }
 
+// A given incidence has been added or modified locally.
+// To upsync the change, we need to construct the .ics data to upload to server.
+// Since the incidence may be an occurrence or recurring series incidence,
+// we cannot simply convert the incidence to iCal data, but instead we have to
+// upsync an .ics containing the whole recurring series.
+QString NotebookSyncAgent::constructLocalChangeIcs(KCalCore::Incidence::Ptr updatedIncidence)
+{
+    // create an in-memory calendar
+    // add to it the required incidences (ie, check if has recurrenceId -> load parent and all instances; etc)
+    // for each of those, we need to do the IncidenceToExport() modifications first
+    // then, export from that calendar to .ics file.
+    KCalCore::MemoryCalendar::Ptr memoryCalendar(new KCalCore::MemoryCalendar(KDateTime::UTC));
+    if (updatedIncidence->hasRecurrenceId() || updatedIncidence->recurs()) {
```

Comment:
User creates recurring event with an exception, then makes the main event non-recurring, then syncs. In case main event is processed first, this could skip the exception?

Then again from spec point of view I'm not sure what should happen. Recurrence id should point to occurrence, but in this case there would be none. As related detail, unclear to me also what should happen if main event is moved. Should recurrence ids be updated or stay as they were. Currently nemo-qml-plugin-calendar doesn't do anything.